janusgraph-solr/src/main/java/org/janusgraph/diskstorage/solr/SolrIndex.java

```diff
@@ -449,25 +461,29 @@ public void restore(Map<String, Map<String, List<IndexEntry>>> documents, KeyInf
         }
     }
 
-    private void deleteIndividualFieldsFromIndex(String collectionName, String keyIdField, String docId, HashSet<IndexEntry> fieldDeletions) throws SolrServerException, IOException {
-        if (fieldDeletions.isEmpty()) return;
-
-        Map<String, String> fieldDeletes = new HashMap<String, String>(1) {{ put("set", null); }};
-
-        SolrInputDocument doc = new SolrInputDocument();
-        doc.addField(keyIdField, docId);
-        StringBuilder sb = new StringBuilder();
-        for (IndexEntry fieldToDelete : fieldDeletions) {
-            doc.addField(fieldToDelete.field, fieldDeletes);
-            sb.append(fieldToDelete).append(",");
+    private Map<String,Object> buildFieldAdditions(List<IndexEntry> content, String collectionName, KeyInformation.IndexRetriever informations) throws BackendException {
+        Map<String,Object> docs = new HashMap<String,Object>();
+        for (IndexEntry addition: content) {
+            KeyInformation keyInformation = informations.get(collectionName, addition.field);
+            switch (keyInformation.getCardinality()) {
+                case SINGLE:
+                    docs.put(addition.field, convertValue(addition.value));
+                    break;
+                case SET:
+                    if (!docs.containsKey(addition.field)) {
+                        docs.put(addition.field, new HashSet<Object>());
+                    }
+                    ((Set<Object>) docs.get(addition.field)).add(convertValue(addition.value));
+                    break;
+                case LIST:
+                    if (!docs.containsKey(addition.field)) {
+                        docs.put(addition.field,  new ArrayList<Object>());
+                    }
+                    ((List<Object>) docs.get(addition.field)).add(convertValue(addition.value));
+                    break;
+            }
         }
-
-        if (logger.isTraceEnabled())
-            logger.trace("Deleting individual fields [{}] for document {}", sb.toString(), docId);
-
-        UpdateRequest singleDocument = newUpdateRequest();
-        singleDocument.add(doc);
-        solrClient.request(singleDocument, collectionName);
+        return Collections.unmodifiableMap(docs);
```

Comment:
You can revert this change as well ... unmodifiable not necessary since method is private and result is used internally.